rust/src/mqtt/parser.rs

```diff
@@ -742,4 +742,386 @@ mod tests {
+            0x75, 0x73, 0x65, 0x72, /* User Name: user */
+            0x00, 0x04, /* Password Length: 4 */
+            0x71, 0x61, 0x71, 0x73, /* Password: pass */
+        ];
+
+        let result = parse_connect(&buf);
+        match result {
+            Ok((remainder, message)) => {
+                assert_eq!(message.protocol_string, "MQTT");
+                assert_eq!(message.protocol_version, 5);
+                assert_eq!(message.username_flag, true);
+                assert_eq!(message.password_flag, true);
+                assert_eq!(message.will_retain, false);
+                assert_eq!(message.will_qos, 0);
+                assert_eq!(message.will_flag, false);
+                assert_eq!(message.clean_session, true);
+                assert_eq!(message.keepalive, 60);
+                assert_eq!(remainder.len(), 0);
+            }
+            Err(Err::Incomplete(_)) => {
+                panic!("Result should not have been incomplete.");
+            }
+            Err(Err::Error(err)) | Err(Err::Failure(err)) => {
+                panic!("Result should not be an error: {:?}.", err);
+            }
+        }
+    }
+
+    #[test]
+    fn test_parse_connack() {
+        let buf = [
+            0x00, /* Acknowledge Flags: 0x00 (0000 000. = Reserved: Not set )(.... ...0 = Session Present: Not set) */
+            0x00, /* Reason Code: Success (0) */
+            0x2f, /* Total Length: 47 */
+            0x22, /* ID: Topic Alias Maximum (0x22) */
+            0x00, 0x0a, /* Value: 10 */
+            0x12, /* ID: Assigned Client Identifier (0x12) */
+            0x00, 0x29, /* Length: 41 */
+            0x61, 0x75, 0x74, 0x6f, 0x2d, 0x31, 0x42, 0x34, 0x33, 0x45, 0x38, 0x30, 0x30, 0x2d,
+            0x30, 0x38, 0x45, 0x33, 0x2d, 0x33, 0x42, 0x41, 0x31, 0x2d, 0x32, 0x45, 0x39, 0x37,
+            0x2d, 0x45, 0x39, 0x41, 0x30, 0x42, 0x34, 0x30, 0x36, 0x34, 0x42, 0x46,
+            0x35, /* 41 byte Value: auto-1B43E800-08E3-3BA1-2E97-E9A0B4064BF5 */
+        ];
+        let client_identifier = "auto-1B43E800-08E3-3BA1-2E97-E9A0B4064BF5";
+
+        let result = parse_connack(5);
+        let input = result(&buf);
+        match input {
+            Ok((remainder, message)) => {
+                let props = message.properties.unwrap();
+                assert_eq!(props[0], MQTTProperty::TOPIC_ALIAS_MAXIMUM(10));
+                assert_eq!(
+                    props[1],
+                    MQTTProperty::ASSIGNED_CLIENT_IDENTIFIER(client_identifier.to_string())
+                );
+                assert_eq!(message.return_code, 0);
+                assert_eq!(message.session_present, false);
+                assert_eq!(remainder.len(), 0);
+            }
+
+            Err(Err::Incomplete(_)) => {
+                panic!("Result should not have been incomplete.");
+            }
+            Err(Err::Error(err)) | Err(Err::Failure(err)) => {
+                panic!("Result should not be an error: {:?}.", err);
+            }
```

Comment:
We could test that there shouldn't be any error as well. e.g. `assert_eq!(input.is_ok(), True)` instead of trying to panic on the errors.
This in not wrong btw.

Reply:
Ok thanks for mentioning this. Should I try doing that in some of the tests or this is fine for now!

Reply:
Seems ok to me. You could use that pattern in future to make the tests more compact 😉